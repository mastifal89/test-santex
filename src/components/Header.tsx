import { useContext, useEffect } from 'react';
import styled from 'styled-components';
import { Cart } from '@styled-icons/bootstrap/Cart';
import { CartContext } from '../hooks/cartContext';
import { types } from '../types';

const CartIcon = styled(Cart)`
  color: white;
  width: 2rem;
  height: 2rem;
  margin-right: 1rem;
`;

const Badge = styled.span`
  background-color: red;
  border-radius: 10%;
  color: white;
  font-size: 0.8rem;
  font-weight: 500;
  padding: 0.3rem;
  position: absolute;
  top: 0.8rem;
  right: 1rem;
`;

const HeaderWrapper = styled.header`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.5);
`;

const Row = styled.div`
  display: flex;
  direction: row;
  justify-content: space-between;
  margin-bottom: 1rem;
`;

const Col = styled.div`
  flex: 1;
`;

export const Header = () => {

  const { cart, dispatch } = useContext(CartContext);

  useEffect(() => {
    if(cart?.cart?.length === 0) {
      dispatch({ type: types.addItemToCart, payload: JSON.parse(localStorage.getItem('cart') || '[]') });
    }
  }, [cart?.cart?.length, dispatch]);

  return (
    <HeaderWrapper style={{ background: 'black' }}>
      <Row>
        <Col>
          <img
            src="https://santex.wpengine.com/wp-content/uploads/2019/02/logo-santex@3x.png"
            alt="logo"
            style={{ marginTop: 15, marginLeft: 10 }}
          />
        </Col>
        <Col
          style={{
            display: 'flex',
            justifyContent: 'flex-end',
            alignItems: 'center',
            marginRight: 10,
            marginTop: 10,
          }}
        >
          <div style={{ color: 'white', marginRight: '1rem' }}>
            {cart && '$ '}{cart && cart.cart?.total}
          </div>
          <CartIcon />
          {cart?.cart?.totalQuantity && <Badge>{cart ? cart.cart?.totalQuantity : null}</Badge>}
        </Col>
      </Row>
    </HeaderWrapper>
  );
};
