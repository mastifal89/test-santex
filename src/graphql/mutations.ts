import { gql } from '@apollo/client';

const ORDER_FRAGMENT = gql`
  fragment ActiveOrder on Order {
    id
    code
    state
    total
    currencyCode
    lines {
      id
      productVariant {
        id
        name
        currencyCode
      }
      unitPriceWithTax
      quantity
      totalPrice
      featuredAsset {
        id
        preview
      }
    }
  }
`;

export const ADD_TO_CART = gql`
    mutation addItemToOrder($productVariantId: ID!, $quantity: Int!) {
        addItemToOrder(productVariantId: $productVariantId, quantity: $quantity) {
            ... on Order {
                id
                total
                subTotal
                totalQuantity
            }
        }
    }
`;

export const GET_ACTIVE_ORDER = gql`
  {
    activeOrder {
      ...ActiveOrder
    }
  }
  ${ORDER_FRAGMENT}
`;