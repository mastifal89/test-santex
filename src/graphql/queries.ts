import { gql } from '@apollo/client';

export const GET_PRODUCTS = gql`
    query getProducts {
        products {
            items {
                id
                name
                description
                assets {
                    source
                }
                variants {
                    id
                    price
                }
            }
        }
    }
`;

export const GET_ORDERS = gql`
    query activeOrder {
        activeOrder {
            id
            code
            state
            total
            currencyCode
            totalQuantity
            lines {
                id
                productVariant {
                    id
                    name
                    currencyCode
                }
            }
        }
    }
`;
