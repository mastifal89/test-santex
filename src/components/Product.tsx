import { useMutation } from '@apollo/client';
import { FC, useContext } from 'react';
import styled from 'styled-components';
import { ADD_TO_CART } from '../graphql/mutations';
import Spinner from './StyledComponents/Spinner';
import { CartContext } from '../hooks/cartContext';
import { types } from '../types';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 350px;
  height: 550px;
  background-color: grey;
  border-radius: 4px;
  padding: 1rem;
  margin: 1rem;

  &:hover {
    opacity: 0.5;
  }
`;

const Title = styled.h1`
  font-size: 2rem;
  font-weight: 300;
  margin: 1rem;
  color: white;
`;

const Description = styled.p`
  color: white;
  font-size: 0.8rem;
  justify-content: flex-end;
`;

const Price = styled.p`
  color: white;
  font-size: 1.2rem;
  font-weight: 500;
  margin: 0 auto;
`;

const Button = styled.button`
  background-color: #ffa500;
  border: none;
  border-radius: 4px;
  color: white;
  font-size: 1rem;
  font-weight: 300;
  padding: 0.5rem;
  margin: 1rem;
  cursor: pointer;
`;

const Image = styled.img`
  max-height: 250px;
`

interface IProps {
  id: number;
  name: string;
  description: string;
  price: number;
  url: string;
}

export const Product: FC<IProps> = ({
  id,
  name,
  description,
  price,
  url,
}): JSX.Element => {

  const { cart, dispatch } = useContext(CartContext);

  const [addItemToOrder, { loading, error }] = useMutation(ADD_TO_CART);

  const addToCart = async (product: any) => {
    
    const { data } = await addItemToOrder({
      variables: {
        productVariantId: product,
        quantity: 1,
      },
    });
    if(!error) {
      dispatch({
        type: types.addItemToCart,
        payload: data.addItemToOrder,
      });
    }
  }

  return (
    <Wrapper>
      <Title>{name}</Title>
      <Description>{description}</Description>
      <Image alt={name} src={url}/>
      <Price>${price}</Price>
      <Button disabled={loading} type='button' onClick={() => addToCart(id)}>{loading ? <div><Spinner />"Loading..."</div> : "Buy"}</Button>
    </Wrapper>
  );
};
