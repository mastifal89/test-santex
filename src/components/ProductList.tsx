import { useQuery } from '@apollo/client';
import { GET_PRODUCTS } from '../graphql/queries';
import { Product } from './Product';
import styled from 'styled-components';
import Spinner from './StyledComponents/Spinner';

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-top: 7rem;
`;

export const ProductList = () => {
  const { loading, error, data } = useQuery(GET_PRODUCTS);

  return (
    <div>
      {loading && <div><Spinner/>Loading...</div>}
      {error && <p>Error :(</p>}
      <Wrapper>
        {data &&
          data.products.items.map((product: any) => (
            <Product
              key={product.id}
              id={product.variants[0]?.id}
              name={product.name}
              description={product.description}
              price={product.variants[0]?.price}
              url={product.assets[0].source}
            />
          ))}
      </Wrapper>
    </div>
  );
}
