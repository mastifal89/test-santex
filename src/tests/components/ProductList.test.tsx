import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { ProductList } from '../../components/ProductList';
import { GET_PRODUCTS } from '../../graphql/queries';

const mocks: any = [];

const mocksWithData: any = [
    {
        request: {
            query: GET_PRODUCTS,
        },
        result: {
            data: {
                products: {
                    items: [
                        {
                            id: '1',
                            name: 'Product 1',
                            description: 'Description 1',
                            variants: [
                                {
                                    id: '1',
                                    price: 1,
                                },
                            ],
                            assets: [
                                {
                                    source: 'https://via.placeholder.com/150',
                                },
                            ],
                        },
                    ],
                },
            },
        },
    },
];

it('should render ProductList correctly', () => {
    const wrapper = TestRenderer.create(
        <MockedProvider mocks={mocks} addTypename={false}>
            <ProductList />
        </MockedProvider>
    );
    expect(wrapper).toMatchSnapshot();
});

it('should render ProductList correctly with data', () => {
    const wrapper = TestRenderer.create(
        <MockedProvider mocks={mocksWithData} addTypename={false}>
            <ProductList />
        </MockedProvider>
    );
    expect(wrapper).toMatchSnapshot();
});