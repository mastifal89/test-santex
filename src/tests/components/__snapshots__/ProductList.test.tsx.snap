// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProductList correctly 1`] = `
<div>
  <div>
    <div
      className="sc-bczRLJ bbIYkx"
    />
    Loading...
  </div>
  <div
    className="sc-iBkjds ikQqWV"
  />
</div>
`;

exports[`should render ProductList correctly with data 1`] = `
<div>
  <div>
    <div
      className="sc-bczRLJ bbIYkx"
    />
    Loading...
  </div>
  <div
    className="sc-iBkjds ikQqWV"
  />
</div>
`;
